control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs

```diff
@@ -0,0 +1,161 @@
+}
+impl NodeNexusReconciler {
+    /// Return a new `Self`.
+    pub(super) fn new() -> Self {
+        Self {
+            counter: PollTimer::from(1), // sets the reconciler polling rate
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for NodeNexusReconciler {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let nodes = context.specs().get_nodes();
+        let mut results = Vec::with_capacity(nodes.len());
+
+        for node in nodes {
+            results.push(check_and_drain_node(context, &node).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+
+    async fn poll_event(&mut self, context: &PollContext) -> bool {
+        matches!(context.event(), PollEvent::TimedRun)
+    }
+}
+
+/// Republish the specified volume on any node other than its current node.
+async fn republish_volume(
+    volume: &mut OperationGuard<ResourceMutex<VolumeSpec>, VolumeSpec>,
+    context: &PollContext,
+    vol_uuid: &VolumeId,
+    frontend_node: &NodeId,
+) -> Result<(), SvcError> {
+    let request = RepublishVolume::new(
+        vol_uuid.clone(),
+        None,
+        frontend_node.clone(),
+        VolumeShareProtocol::Nvmf,
+        false,
+    );
+    tracing::info!(
+        volume.uuid = vol_uuid.as_str(),
+        "Attempting to republish volume"
+    );
+
+    volume.republish(context.registry(), &request).await?;
+
+    Ok(())
+}
+
+/// Drain the specified node if in draining state
+async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> PollResult {
+    if !node_spec.is_draining() {
+        return PollResult::Ok(PollerState::Idle);
+    }
+
+    let node_id = node_spec.id();
+    tracing::trace!(node.id = node_spec.id().as_str(), "Draining node");
+    let vol_specs = context.specs().get_locked_volumes();
+
+    let mut move_failures = false;
+
+    // Iterate through all the volumes, find those with a nexus hosted on the
+    // node and move each one away.
+    for vol_spec in vol_specs {
+        match vol_spec.operation_guard() {
+            Ok(mut guarded_vol_spec) => {
+                if let Some(target) = guarded_vol_spec.as_ref().target() {
+                    if target.node() != node_id {
+                        continue; // some other node's volume, ignore
+                    }
+                    let nexus_id = target.nexus().clone();
+                    let vol_id = guarded_vol_spec.as_ref().uuid.clone();
+
+                    let config = guarded_vol_spec.as_ref().config().clone();
+                    if config.is_none() {
+                        tracing::error!("Failed to get config");
+                        move_failures = true;
+                        continue;
+                    }
+                    let frontend_nodes = config.unwrap().frontend().node_names();
+                    if frontend_nodes.is_empty() {
```

Comment:
Maybe just pass "" in this case?
`let frontend_node = config.frontend().node_names().first().unwrap_or_default();`
Otherwise the node will never be drained?

Reply:
Because "" is an acceptable value for the frontend node?
Got it.

Reply:
IIRC "" was being used on some tests (I then changed it to a proper name during code refactoring), so I guess it's acceptable? Not sure then, might be worth checking

Reply:
As discussed, the republish uses "" via String::default() if the frontend node isn't defined.